packages/pkl.lua/lua.pkl

```diff
@@ -0,0 +1,292 @@
+      else p.first == p.second
+    )
+
+  local function findTypedConverter(klass: reflect.Class?): ((unknown) -> Any)? =
+    if (klass == null) null else
+    converters.getOrNull(klass.reflectee) ??
+    findTypedConverter(klass.superclass)
+
+  // ## Rendering functions ##
+
+  local function render(value: Any, path: List<PathEntry>?, level: UInt?): String =
+    if (value is String) renderString(value, level != null)
+    else if (value is Boolean) value.toString()
+    else if (value is Number) renderNumber(value)
+    else if (value is Null) "nil"
+    else if (value is Mapping|Map) renderMap(value, path, level)
+    else if (value is Listing|List) renderList(value, path, level)
+    else if (value is Dynamic) renderDynamic(value, path, level)
+    else if (value is RenderDirective) value.text
+    else if (value is LuaRenderDirective) "\(value.before ?? "")\(render(value.value, path, level))\(value.after ?? "")"
+    else if (value is Typed) renderDynamic(value, path, level)
+    else throw("Cannot render value of type `\(value.getClass())` as Lua.\nValue: \(value)")
+
+  local function renderKey(key: Any, prefix: String): String =
+    if (key is Null) throw("Lua table keys cannot be null")
+    else if (key is Number && key.isNaN) throw("Lua table keys cannot be NaN")
+    else if (key is LuaIdentifier) key
+    else if (key is RenderDirective) key.text
+    else "\(prefix)[\(render(key, null, null))]"
+
+  local function renderString(value: String, multiline: Boolean): String =
+    let (delim = if (value.contains("\"") && !value.contains("'")) "'" else "\"")
+    if (multiline && value.contains(Regex(#"[^\n]\n++[^\n]"#)))
+      // there are interior newlines, we'll do a multiline style
+      if (value.contains(Regex(#"[\p{Cntrl}\#(delim)&&[^\n ]]"#)))
+        // we need escapes, we'll do multiline with escaped newlines
+        "\(delim)\(escapeString(value, delim, true))\(delim)"
+      else
+        IntSeq(0, 10)
+        .map((n) -> "=".repeat(n))
+        .findOrNull((eq) -> !value.contains("]\(eq)]"))
+        .ifNonNull((eq) -> "[\(eq)[\n\(value)]\(eq)]")
+        // if we can't find a good ]==] ending, use the other style
+        ?? "\(delim)\(escapeString(value, delim, true))\(delim)"
+    else "\(delim)\(escapeString(value, delim, false))\(delim)"
+
+  local function escapeString(value: String, delim: String, multiline: Boolean): String =
+    value.replaceAllMapped(Regex(#"[\p{Cntrl}\\\#(delim)&&[^ ]]"#), (match) ->
+      if (match.value == "\u{7}") "\\a"
+      else if (match.value == "\u{8}") "\\b"
+      else if (match.value == "\u{12}") "\\f"
+      else if (match.value == "\n") if (multiline) "\\\n" else "\\n"
+      else if (match.value == "\r") "\\r"
+      else if (match.value == "\t") "\\t"
+      else if (match.value == "\u{11}") "\\v"
+      else if (match.value == delim) "\\\(delim)"
+      else "\\u{\(match.value.codepoints.first.toRadixString(16))}")
```

Comment:
This would be a runtime error, so I guess this isn't being tested right now.

```suggestion
      else "\\u{\(match.value.codePoints.first.toRadixString(16))}")
```

Also, `codePoints.first` is a little suspicious. Do you need to worry about surrogate pairs here?

Reply:
Good catch! I'll make sure there's a test that exercises this code path.